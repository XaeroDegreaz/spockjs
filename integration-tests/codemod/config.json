{
  "plugins": [
    [
      "@spockjs/babel-plugin-spock",
      {
        "powerAssert": false,
        "autoImport": "assert",
        "assertFunctionName": "assert"
      }
    ]
  ]
}
