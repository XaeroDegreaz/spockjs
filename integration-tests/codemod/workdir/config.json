{
  "plugins": [
    [
      "@spockjs/babel-plugin-spock",
      {
        "powerAssert": false,
        "autoImport": "assert",
        "assertFunctionName": "assert",
        "presets": ["@spockjs/preset-sinon-mocks"]
      }
    ]
  ]
}
