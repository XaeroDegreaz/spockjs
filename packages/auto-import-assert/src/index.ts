import { Scope } from '@babel/traverse';
import * as BabelTypes from '@babel/types';

export const findExistingImport = (
  scope: Scope,
  t: typeof BabelTypes,
  source: string,
) => {
  const program = scope.getProgramParent().path.node as BabelTypes.Program;

  // try to find existing default import from source
  for (const stmt of program.body) {
    if (t.isImportDeclaration(stmt) && stmt.source.value === source) {
      const defaultSpecifier = stmt.specifiers.find(specifier =>
        t.isImportDefaultSpecifier(specifier),
      ) as BabelTypes.ImportDefaultSpecifier | undefined;
      if (defaultSpecifier) {
        const { local } = defaultSpecifier;
        const { name } = local;
        // make sure the import is not shadowed in the scope
        if (scope.bindingIdentifierEquals(name, local)) {
          return name;
        }
      }
    }
  }
  return;
};

export const addImport = (
  scope: Scope,
  t: typeof BabelTypes,
  source: string,
  exactName: string,
  fallbackNameHint: string,
) => {
  if (exactName) {
    // clear possible binding conflicts
    scope.rename(exactName);
  }

  const program = scope.getProgramParent().path;

  // Generate default import from exact name or default name hint.
  // Using @babel/helper-module-imports would give us commonjs support
  // for free without requiring @babel/preset-env, however
  // 1. it does not support exact names, so we would still need to
  //    use our manual import generation if assertFunctionName is set and
  // 2. it might generate SequenceExpressions instead of Identifiers,
  //    which are annoying to deal with because of powerAssert patterns.
  const id = exactName
    ? t.identifier(exactName)
    : scope.generateUidIdentifier(fallbackNameHint);
  (program as any).unshiftContainer(
    'body',
    t.importDeclaration(
      [t.importDefaultSpecifier(id)],
      t.stringLiteral(source),
    ),
  );

  return id.name;
};
