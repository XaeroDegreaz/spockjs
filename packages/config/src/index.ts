import * as t from 'io-ts';
import { failure } from 'io-ts/lib/PathReporter';

import { Hooks } from './hooks';

/**
 * Config as it can be provided by the user.
 * Comments indicate transformations made to conform to InternalConfig format.
 */
export const configType = t.exact(
  t.partial({
    powerAssert: t.boolean,
    autoImport: t.union([t.boolean, t.string]),
    staticTruthCheck: t.boolean,
    assertFunctionName: t.string,
    presets: t.array(t.string),
  }),
  'Config',
);
export type Config = t.TypeOf<typeof configType>;

/**
 * Config that disables all but the core set of features.
 */
export const minimalConfig: Config = {
  powerAssert: false,
  autoImport: false,
};

/**
 * Config containing the settings used if they are missing in the user-specified config.
 */
export const defaultConfig = {
  powerAssert: true,
  autoImport: true,
  staticTruthCheck: false,
  assertFunctionName: '',
  presets: [],
};

/**
 * Config after sanitizing in extractConfigFromState.
 */
export interface InternalConfig {
  readonly powerAssert: boolean;
  readonly autoImport: string; // empty => no auto import
  readonly staticTruthCheck: boolean;
  readonly assertFunctionName: string; // empty => no custom assert function name
  readonly hooks: Hooks;
}

export const extractConfigFromState = ({
  opts,
}: {
  opts: any;
}): InternalConfig => {
  const decoded = configType.decode(opts);

  let {
    powerAssert,
    autoImport,
    staticTruthCheck,
    assertFunctionName,
    presets,
  } = decoded.getOrElseL(errors => {
    throw new Error(
      ['Invalid plugin configuration:', ...failure(errors)].join('\n'),
    );
  });

  // powerAssert
  if (powerAssert === undefined) {
    ({ powerAssert } = defaultConfig);
  }

  // autoImport
  if (autoImport === undefined) {
    ({ autoImport } = defaultConfig);
  }
  if (autoImport === false) {
    autoImport = '';
  }
  if (autoImport === true) {
    autoImport = 'power-assert';
  }

  // staticTruthCheck
  if (staticTruthCheck === undefined) {
    ({ staticTruthCheck } = defaultConfig);
  }

  // assertFunctionName
  if (assertFunctionName === undefined) {
    ({ assertFunctionName } = defaultConfig);
  }

  // presets
  if (presets === undefined) {
    ({ presets } = defaultConfig);
  }
  const hooks = presets.map(preset => require(preset) as Hooks).reduce(
    (accHooks, moreHooks) => ({
      assertionPostProcessors: [
        ...accHooks.assertionPostProcessors,
        ...moreHooks.assertionPostProcessors,
      ],
    }),
    { assertionPostProcessors: [] },
  );

  return {
    powerAssert,
    autoImport,
    staticTruthCheck,
    assertFunctionName,
    hooks,
  };
};
