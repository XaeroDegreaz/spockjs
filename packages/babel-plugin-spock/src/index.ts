import { PluginObj } from '@babel/core';
import { NodePath } from '@babel/traverse';
import * as BabelTypes from '@babel/types';

import { extractConfigFromState } from '@spockjs/config';

import assertifyStatement from './assertify-statement';

const assertionBlockLabels = ['expect', 'then'];

const plugin = (babel: { types: typeof BabelTypes }): PluginObj => {
  return {
    visitor: {
      LabeledStatement(path, state) {
        const config = extractConfigFromState(state);
        const assertify = assertifyStatement(babel, state, config);

        if (assertionBlockLabels.includes(path.node.label.name)) {
          const bodyPath = path.get('body') as NodePath<BabelTypes.Statement>;
          switch (bodyPath.type) {
            case 'BlockStatement':
              // power-assert may add statements in betweeen,
              // so never reuse body array
              const statementPaths = () =>
                (bodyPath.get('body') as any) as NodePath<
                  BabelTypes.Statement
                >[];

              statementPaths().forEach(assertify);
              path.replaceWithMultiple(
                statementPaths().map(stmtPath => stmtPath.node),
              );
              break;
            default:
              assertify(bodyPath);
              path.replaceWith(bodyPath);
          }
        }
      },
    },
  };
};

export default plugin;
