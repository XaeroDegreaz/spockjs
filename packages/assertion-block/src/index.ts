import print from '@babel/generator';
import { NodePath } from '@babel/traverse';
import * as BabelTypes from '@babel/types';

import { addImport, findExistingImport } from '@spockjs/auto-import-assert';
import { InternalConfig } from '@spockjs/config';
import empowerAssert from '@spockjs/power-assert';
import checkStatically from '@spockjs/static-check';

export default (
  babel: { types: typeof BabelTypes },
  state: any,
  config: InternalConfig,
) => {
  const { types: t } = babel;
  const {
    assertFunctionName,
    autoImport,
    powerAssert,
    staticTruthCheck,
    hooks: { assertionPostProcessors },
  } = config;
  const postProcessors = assertionPostProcessors.map(processor =>
    processor(t, config),
  );

  return (statementPath: NodePath<BabelTypes.Statement>) => {
    if (statementPath.isExpressionStatement()) {
      const { scope, node: statement } = statementPath;

      if (staticTruthCheck) {
        checkStatically(statementPath.get('expression') as NodePath<
          BabelTypes.Expression
        >);
      }

      const origExpr = statement.expression;

      const assertIdentifier = t.identifier(
        autoImport
          ? // reuse or add import from given source
            findExistingImport(scope, t, autoImport) ||
            addImport(scope, t, autoImport, assertFunctionName, 'assert')
          : // assume assert identifier is already available
            assertFunctionName || 'assert',
      );
      assertIdentifier.loc = {
        start: origExpr.loc.start,
        end: origExpr.loc.start,
      };

      const newExpr = t.callExpression(assertIdentifier, [
        origExpr,
        t.stringLiteral(`(${print(origExpr).code}) is not truthy`),
      ]);
      newExpr.loc = origExpr.loc;
      statement.expression = newExpr;

      // register added import with scope so it is found for the next assertion
      // for some reason, this needs to happen after we have referenced the import
      // with our call expression, otherwise the import will be removed
      (scope.getProgramParent() as any).crawl();

      const {
        path: processedExpressionStatementPath,
        patterns,
      } = postProcessors.reduce(
        ({ path, patterns }, postProcess) => postProcess(path, patterns),
        {
          path: statementPath,
          patterns: [`${assertIdentifier.name}(value, [message])`],
        },
      );

      if (powerAssert) {
        empowerAssert(babel, state, patterns, processedExpressionStatementPath);
      }
    } else {
      throw statementPath.buildCodeFrameError(
        `Expected an expression statement, but got a statement of type ${
          statementPath.type
        }`,
      );
    }
  };
};

export const labels = ['expect', 'then'];
