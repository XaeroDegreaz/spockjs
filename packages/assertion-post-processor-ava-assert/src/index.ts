import { CallExpression } from '@babel/types';

import { AssertionPostProcessor } from '@spockjs/config/src/hooks';

const processor: AssertionPostProcessor = (t, config, path) => {
  const assertionExpression = path.node.expression as CallExpression;

  const tTruthy = t.memberExpression(t.identifier('t'), t.identifier('truthy'));
  tTruthy.loc = tTruthy.object.loc = tTruthy.property.loc =
    assertionExpression.callee.loc;
  // When espower "purifies" ASTs internally to match against the patterns,
  // our MemberExpression retains an `optional: undefined` property,
  // which the pattern does not have, preventing it from being matched.
  delete (tTruthy as any).optional;

  assertionExpression.callee = tTruthy;
  return { path, patterns: ['t.truthy(value)'] };
};

export default processor;

// TODO func test
