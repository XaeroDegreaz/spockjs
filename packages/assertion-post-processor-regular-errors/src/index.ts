import { addNamed } from '@babel/helper-module-imports';
import template from '@babel/template';
import { NodePath } from '@babel/traverse';
import { ExpressionStatement, TryStatement } from '@babel/types';

import { AssertionPostProcessor } from '@spockjs/config/src/hooks';

import autoImportDisabled from './auto-import-disabled-warning';

const createTryStatement = template(`
  try {
    ASSERTION
  } catch(e) {
    if (e instanceof ASSERTION_ERROR) {
      throw new Error(e.message);
    }
    throw e;
  }
`);

const processor: AssertionPostProcessor = (t, { autoImport }, originalPath) => {
  const assertionErrorImportedName = 'AssertionError';
  const assertionErrorLocal = autoImport
    ? addNamed(originalPath, assertionErrorImportedName, autoImport)
    : t.identifier(assertionErrorImportedName);
  if (!autoImport && !autoImportDisabled.warned) {
    console.warn(
      'Generating catch clauses to intercept AssertionErrors ' +
        'with autoImport disabled. The AssertionError class must be imported ' +
        `and available in scope as \`${assertionErrorImportedName}\``,
    );
    autoImportDisabled.warned = true;
  }

  const { node: assertionNode } = originalPath;
  const tryStatement = createTryStatement({
    ASSERTION: assertionNode,
    ASSERTION_ERROR: assertionErrorLocal,
  }) as TryStatement;

  originalPath.replaceWith(tryStatement);
  const tryPath = (originalPath as any) as NodePath<TryStatement>;

  // set some loc info so test runners can tell the user where an error comes from
  tryPath.traverse({
    ThrowStatement(path) {
      path.node.loc = {
        start: assertionNode.loc.end,
        end: assertionNode.loc.end,
      };
    },
  });

  // return the original assertion ExpressionStatement
  return tryPath.get('block.body.0') as NodePath<ExpressionStatement>;
};

export default processor;
