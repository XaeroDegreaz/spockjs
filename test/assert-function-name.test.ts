import { transform } from '@babel/core';

import plugin from '../src';
import { Config, minimalConfig } from '../src/config';

test('generates specified assert function', () => {
  const { code } = transform(`expect: 1 === 1;`, {
    plugins: [
      [plugin, { ...minimalConfig, assertFunctionName: 'check' } as Config],
    ],
  });
  expect(code).toMatchSnapshot();
});

test('generates import with specified name when autoImport is set', () => {
  const { code } = transform(`expect: 1 === 1;`, {
    plugins: [
      [
        plugin,
        {
          ...minimalConfig,
          assertFunctionName: 'check',
          autoImport: true,
        } as Config,
      ],
    ],
  });
  expect(code).toMatchSnapshot();
});
