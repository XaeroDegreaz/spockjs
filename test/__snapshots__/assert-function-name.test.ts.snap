// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates import with specified name when autoImport is set 1`] = `
"import check from \\"power-assert\\";

expect: check(1 === 1);"
`;

exports[`generates specified assert function 1`] = `"expect: check(1 === 1);"`;
